File: Ruby/hh-autopay-ruby/helpers/z_crm.rb
class CrmApi
  attr_accessor :retrieval_time, :conn

  def initialize
    self.conn = Faraday.new 'https://www.zohoapis.com/crm/v3/'
    self.refresh_token
    self
  end
  
  def refresh_token
    ref_url = "https://accounts.zoho.com/oauth/v2/token?refresh_token=#{ENV['REFRESH_TOKEN']}&client_id=#{ENV['CLIENT_ID']}&client_secret=#{ENV['CLIENT_SECRET']}&grant_type=refresh_token"
    response = @conn.post(ref_url)
    token = JSON.parse(response.body)
    
    if token && response.status == 200
      $logger.info('Zoho refresh Token received')
      @conn.headers['Authorization'] = "Zoho-oauthtoken #{token['access_token']}"
    else
      $logger.error('Zoho refresh Token not received')
      $logger.debug(response.body)
      $logger.debug(response.status)
    end
    self.retrieval_time = Time.now.strftime "%H:%M"
  end

  def search(z_module, params)
    url = "#{z_module}/search?criteria=#{params}"
    response = @conn.get(url,nil)
    if response == 200
      data = JSON.parse(response.body)
      $logger.info('Zoho search record returned data')
      return data['data']
    else
      $logger.info('Zoho search record returned no data')
      $logger.debug(response.body)
      $logger.debug(response.status)
      return []
    end
  end

  def post_put(z_module, payload, post_type='')
    api_url = "#{z_module}#{post_type}"
    payload_length = payload.length
    num_of_req = (payload_length.to_f / 100 ).ceil 
    start_ind = 00
    end_ind =
      if payload_length > 100
        99
      else
        payload_length - 1
      end
    resp_array = []
    for iteration in 1..num_of_req do
      data_hash = {data: payload[start_ind..end_ind]}
      post_resp = post(api_url, data_hash)
      resp_array.append(post_resp)
      start_ind += 100
      end_ind += 100
      next_iter = iteration + 1
      end_ind = payload_len - 1 if next_iter == num_of_req
    end
    resp_array
  end

  def post(url, payload)
    response = @conn.post(url, payload.to_json)
    case response.status
    when 200, 201, 202, 204
      data = JSON.parse(response.body)
      $logger.info('Zoho post successful')
      return data['data']
    else
      $logger.error('Zoho post failed')
      $logger.debug(response.body)
      $logger.debug(response.status)
      return 'fail'
    end
  end
end

File: hh-autopay-ruby/functions/z_crm.js
const axios = require('axios');
const logger = require('./logger'); // Import the logger module

class CrmApi {
    constructor() {
        this.conn = axios.create({
            baseURL: 'https://www.zohoapis.com/crm/v3/',
            headers: { 'Content-Type': 'application/json' }
        });
        this.refreshToken();
    }

    async refreshToken() {
        try {
            const refUrl = `https://accounts.zoho.com/oauth/v2/token?refresh_token=${process.env.REFRESH_TOKEN}&client_id=${process.env.CLIENT_ID}&client_secret=${process.env.CLIENT_SECRET}&grant_type=refresh_token`;
            const response = await this.conn.post(refUrl);

            const token = response.data;

            if (token && response.status === 200) {
                logger.info('Zoho refresh Token received');
                this.conn.defaults.headers['Authorization'] = `Zoho-oauthtoken ${token.access_token}`;
            } else {
                logger.error('Zoho refresh Token not received');
                logger.debug(response.data);
                logger.debug(response.status);
            }

            this.retrievalTime = new Date().toLocaleTimeString();
        } catch (error) {
            logger.error('Error during Zoho token refresh');
            logger.debug(error);
        }
    }

    async search(zModule, params) {
        try {
            const url = `${zModule}/search?criteria=${params}`;
            const response = await this.conn.get(url);

            if (response.status === 200) {
                logger.info('Zoho search record returned data');
                return response.data.data || [];
            } else {
                logger.info('Zoho search record returned no data');
                logger.debug(response.data);
                logger.debug(response.status);
                return [];
            }
        } catch (error) {
            logger.error('Error during Zoho search');
            logger.debug(error);
            return [];
        }
    }

    async postPut(zModule, payload, postType = '') {
        const apiUrl = `${zModule}${postType}`;
        const payloadLength = payload.length;
        const numOfReq = Math.ceil(payloadLength / 100);
        let startInd = 0;
        let endInd = payloadLength > 100 ? 99 : payloadLength - 1;
        let respArray = [];

        for (let iteration = 1; iteration <= numOfReq; iteration++) {
            const dataHash = { data: payload.slice(startInd, endInd + 1) };
            const postResp = await this.post(apiUrl, dataHash);
            respArray.push(postResp);

            startInd += 100;
            endInd += 100;

            if (iteration === numOfReq) {
                endInd = payloadLength - 1;
            }
        }
        return respArray;
    }

    async post(url, payload) {
        try {
            const response = await this.conn.post(url, payload);

            if ([200, 201, 202, 204].includes(response.status)) {
                logger.info('Zoho post successful');
                return response.data.data;
            } else {
                logger.error('Zoho post failed');
                logger.debug(response.data);
                logger.debug(response.status);
                return 'fail';
            }
        } catch (error) {
            logger.error('Error during Zoho post');
            logger.debug(error);
            return 'fail';
        }
    }
}

module.exports = CrmApi;
